src/access/pg_tde_tdemap.c

```diff
@@ -330,23 +177,445 @@ tde_sprint_masterkey(const keyData *k)
+	 * to be less frequent than vacuum. So let's keep this function here rather
+	 * than overloading the vacuum process.
+	 */
+	while(FileRead(map_file, &map_entry, MAP_ENTRY_SIZE, 0, WAIT_EVENT_DATA_FILE_READ))
+	{
+		/* We found a valid entry for the relNumber */
+		if (map_entry.flags == MAP_ENTRY_VALID && map_entry.relNumber == rlocator->relNumber)
+		{
+			found = true;
+			break;
+		}
+
+		/* Increment the key index only if we didn't get a valid slot. */
+		key_index++;
+	}
+
+	/* Let's close the file. */
+	FileClose(map_file);
+
+	/* An entry was expected but none was found */
+	if (found == false)
+	{
+		ereport(FATAL,
+				(errcode(ERRCODE_NO_DATA_FOUND),
+					errmsg("Could not find the required entry for relNumber %d in tde map file \"%s\": %m",
+						rlocator->relNumber,
+						db_map_path)));
+	}
+
+	/* Return the index */
+	return key_index;
+}
+
+/*
+ * Key Data [pg_tde.dat]:
+ * 		header: {Format Version: x}
+ * 		data: {Encrypted Key}
+ * 
+ * Requires a valid index of the key to be written. The function with seek to
+ * the required location in the file. Any holes will be filled when another
+ * job finds an empty index.
+ */
+void
+pg_tde_write_keydata(char *db_keydata_path, const char *master_key_name, int32 key_index, RelKeysData *enc_rel_key_data)
+{
+	File keydata_file = -1;
+	off_t write_pos;
+	bool is_new_file;
+	size_t key_size;
+
+	/* Open and vaidate file for basic correctness. */
+	keydata_file = pg_tde_open_file(db_keydata_path, master_key_name, O_RDWR | O_CREAT, &is_new_file);
+
+	/* Calculate the writing position in the file. */
+	write_pos = TDE_FILE_HEADER_SIZE + (key_index * INTERNAL_KEY_LEN);
+
+	if (lseek(keydata_file, write_pos, SEEK_SET) == -1)
+	{
+		ereport(FATAL,
+				(errcode_for_file_access(),
+					errmsg("Could not seek in tde key data file \"%s\": %m",
+						db_keydata_path)));
+	}
+
+	key_size = SizeOfRelKeysData(enc_rel_key_data->internal_keys_len);
+
+	if (FileWrite(keydata_file, enc_rel_key_data, key_size, 0, WAIT_EVENT_DATA_FILE_WRITE) != key_size)
+	{
+		ereport(FATAL,
+				(errcode_for_file_access(),
+					errmsg("Could not write tde key data file \"%s\": %m",
+						db_keydata_path)));
+	}
+
+	/* Let's close the file. */
+	FileClose(keydata_file);
+}
+
+/*
+ * 
+ */
+RelKeysData *
+pg_tde_read_keydata(char *db_keydata_path, int32 key_index, const char *master_key_name)
+{
+	File keydata_file = -1;
+	RelKeysData *enc_rel_key_data;
+	off_t read_pos;
+	bool is_new_file;
+	size_t key_size;
+	size_t key_data_size;
+	int key_count = 1;
+
+	/* Open and vaidate file for basic correctness. */
+	keydata_file = pg_tde_open_file(db_keydata_path, master_key_name, O_RDONLY, &is_new_file);
+
+	/*
+	 * Allocate in TopMemoryContext and don't pfree sice we add it to
+	 * the cache as well
+	 */
+	key_size = SizeOfRelKeysData(key_count);
+	key_data_size = key_size - SizeOfRelKeysDataHeader;
+
+	/* Allocate and fill in the structure */
+	enc_rel_key_data = (RelKeysData *) palloc(key_size);
```

Comment:
Comment above says `Allocate in TopMemoryContext` so here should be?
```suggestion
	enc_rel_key_data = (RelKeysData *) MemoryContextAlloc(TopMemoryContext, key_size);
```